File: src/rng/mersennetwister.cpp
/* ----------------------------------------------------------------------------
  This file is a part of the AVBMC library, which follows the GPL-3.0 License.
  For license information, see LICENSE file in the top directory, 
  https://github.com/evenmn/avbmc/LICENSE.

  Author(s): Even M. Nordhagen
  Email(s): evenmn@mn.uio.no
  Date: 2022-06-03 (last changed 2022-06-03)
---------------------------------------------------------------------------- */

/* ----------------------------------------------------------------------------
  Mersenne-Twister pseudo random number generator (prng). The prng has a 
  period of 2^19937, and is popularly abbreviated mt19937. The algorithm was
  first proposed and described by ... This code is just a wrapper around the
  standard C++ implementation of Mersenne-Twister.
---------------------------------------------------------------------------- */

#include <iostream>
#include <random>
#include <vector>
#include <algorithm>
#include <cassert>

#include "mersennetwister.h"


/* ----------------------------------------------------------------------------
   Mersenne Twister constructor
---------------------------------------------------------------------------- */

MersenneTwister::MersenneTwister(int seed_)
    : RandomNumberGenerator(), generator(seed())
{
    label = "Mersenne-Twister";
    if (seed_ > 0) {
        generator.seed (seed_);
    }

    //std::uniform_real_distribution randu_(0, 1);
    //randu = randu_;
}


/* ----------------------------------------------------------------------------
   Set seed
---------------------------------------------------------------------------- */

void MersenneTwister::set_seed(unsigned int seed)
{
    generator.seed (seed);
}


/* ----------------------------------------------------------------------------
   Returns a double drawn from a Gaussian distribution with mean 'mean' and
   variance 'variance'
---------------------------------------------------------------------------- */

double MersenneTwister::next_gaussian(double mean, double variance)
{
    std::normal_distribution<double> dis(mean, variance);
    return dis(generator);
}


/* ----------------------------------------------------------------------------
   Returns an interger between 0 and (including) 'upper_limit'
---------------------------------------------------------------------------- */

int MersenneTwister::next_int(int upper_limit)
{
    std::uniform_int_distribution<int> dis(0, upper_limit - 1);
    return dis(generator);
}


/* ----------------------------------------------------------------------------
   Returns a double between 0 and 1 drawn from a uniform distribution
---------------------------------------------------------------------------- */

double MersenneTwister::next_double()
{
    //std::uniform_real_distribution<double> dis(0, 1);
    //return dis(generator);
    return randu(generator);
}


/* ----------------------------------------------------------------------------
   Inspired by numpy.random.choice, but simplified as we always want to draw
   only one sample. The samples are also always range(len(probabilities), so we
   don't need to take them as an argument.
---------------------------------------------------------------------------- */

int MersenneTwister::choice(const std::vector<double> &probabilities)
{
    double r = next_double();
    double p = 0.;
    for (unsigned int i=0; i<probabilities.size(); i++) {
        p += probabilities[i];
        if (r<=p) return i;
    }
    return 0;
}


/* ----------------------------------------------------------------------------
   Shuffle vector randomly
---------------------------------------------------------------------------- */

std::vector<unsigned int> MersenneTwister::shuffle(std::vector<unsigned int> &vector)
{
    std::shuffle(vector.begin(), vector.end(), generator);
    return vector;
}

File: tests/rng/mersennetwister.py
import avbmc as mc

system = mc.System()

# create rng object
rng = mc.MersenneTwister(seed=12345)
rng = mc.MersenneTwister(seed=-1)
rng = mc.MersenneTwister()

# test RNG
N = 1000
probabilities1 = [1.]
probabilities2 = [0., 0., 1.]
probabilities3 = [0., 0.5, 0., 0.5, 0.]
to_shuffle = [1, 2, 3, 4, 5]
for _ in range(N):
    i = rng.next_int(10)
    assert 0 <= i < 10
    d = rng.next_double()
    assert 0 < d < 1
    d = rng.next_gaussian(0, 0)
    assert (d - 0.0) < 1e-14
    i = rng.choice(probabilities1)
    assert i == 0
    i = rng.choice(probabilities2)
    assert i == 2
    i = rng.choice(probabilities3)
    assert i == 1 or i == 3
    shuffled = rng.shuffle(to_shuffle)
    assert len(shuffled) == len(to_shuffle)

# initialize from object
system.set_rng(rng)
    
# initialize by keyword
system.set_rng("mersennetwister")
system.set_rng("mt19937")

# set seed
system.set_seed(12345)
